### DISCUSSION TITLE
Allocation of cores

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [sybrenjansen/mpire#97] [Q&A]
Hello. I'm new to Mpire but have experience of ipyparallel/MPI and a little with Ray Core. My use case is:

An online service runs on a single multicore machine. The cores are allocated to perform specific functions. For example, for an 8-core machine, core's 1-4 perform function A in their associated worker nodes in parallel. The function A operates on an ML model with the proviso each ML model at each core is different (like an ensemble). A request sent to the cores 1-4 returns results to the master. Also, need to pass data from one group of core's to another group.

Is the above possible with Mpire?

Lastly, can MPI be used to communicate between the worker nodes on a single multicore machine? Thanks

**Comment by sybrenjansen** [comment]
To answer your last question first: no, there's no support for MPI in mpire. 

I think you can get pretty far with mpire. Let me just try to code it (see explanation below):

```
from multiprocessing import Queue
from mpire import WorkerPool

def init_A(worker_id, shared_queues, worker_state):
    # Load a different ML model for each worker
    if worker_id == 0:
        worker_state["model"] = ...
    elif worker_id == 1:
        worker_state["model"] = ...
    else:
        ...

def function_A(worker_id, shared_queues, worker_state, args, **kwargs):
    q_A_to_B, q_B_to_A = shared_queues
    result = worker_state["model"].predict(...)
    q_A_to_B.put(result)   # pass data to other pool
    return result   # return result to main process

def function_B(worker_id, shared_queues, args, **kwargs):
    ...

def main():
    q_A_to_B = Queue()
    q_B_to_A = Queue()
    shared_queues = (q_A_to_B, q_B_to_A)
    with Workerpool(n_jobs=4, cpu_ids=range(4), pass_worker_id=True, use_worker_state=True, order_tasks=True, shared_objects=shared_queues) as pool_A, \
            Workerpool(n_jobs=4, cpu_ids=range(4, 8), pass_worker_id=True, order_tasks=True, shared_objects=shared_queues) as pool_B:

        # Add tasks 
        future_results_A = [pool_A.apply_async(function_A, task_args, task_kwargs, worker_init=init_A) for _ in range(4)]
        future_results_B = [pool_B.apply_async(function_B, task_args, task_kwargs) for _ in range(4)]
```
2 workerpools let's you use different functions on a subset of CPUs. Note that I've added the `cpu_ids` option to pin processes to specific CPUs. Not sure if that was what you were after, but it is there.

I've set `order_tasks=True` so if we add as many tasks to the pool as the number of workers, each worker will get one of the tasks guaranteed.

The combination of `use_worker_state=True`, `pass_worker_id=True`, and the `worker_init` argument will ensure each worker of `pool_A` will be able to load a different ML model and store it in its state. This state can then be used within the function you want to call. The results of the function are stored in `AsyncResult` objects (i.e, the `future_results` objects) and can be accessed in the main process whenever they're ready.

If you want to be able to pass data from one group to another, you can create two additional queues and pass them on as `shared_objects` to both pools. Each worker can then make use of these queues. Note though, that if you send data in this queue from one worker in pool A, only one worker in pool B will be able to get it. If you need to broadcast data, you can create as many queues as there are workers and then you send the data to each queue and let the worker only read from one queue based on the `worker_id`.

So yeah, I believe it's possible.

Let me know if you have any questions

**Reply by dineshbvadhia** [reply]
Apologies for editing the original question mid-stream as I wanted to simplify the use-case. Thank-you for the example code and detailed description - it is appreciated and now have a better understanding of Mpire :)

Cpu pinning is needed. Also, the workers need to stay alive during the service lifespan. Some questions: 

a. With worker_init, how do you send one or more modules?

b. If the 'same' request is sent (broadcast) to every worker (in group A) then should order_tasks=False?

**Reply by sybrenjansen** [reply]
a) I'm not sure what you mean exactly with modules. But the `worker_init` function is only called once per worker and is the same for each worker in the pool. You can change the init per worker in the pool by using an `if/else` switch and using the `worker_id` to determine which worker gets what. This is similar to what I did in my `init_A` example. Once the workers are started changing the `worker_init` function doesn't do anything, as the workers have already started.

b) There's no broadcast functionality currently in mpire. The only way to do that right now is to use `order_tasks=True` and send the same object as many times as there are workers. If you would like broadcast functionality, then please open an issue with how you would want this to behave. E.g., I can imagine having a `pool.broadcast(object, worker_ids=None)` function or something that sends the object to the specified workers (None to specify all). But then, where does the object go? Should it be automatically stored in the `worker_state`? Any suggestions?

**Reply by dineshbvadhia** [reply]
a) worker_state and worker_init
I want to determine if my ipyparallel/mpi code can be converted to Mpire. How do I send a .py file (a module) which includes import statements to each worker to get it up and running (as a keep_alive)) prior to receiving requests from the master? A 'bootstrap' function in the .py file is called from the master and executed on each worker to get it ready. Once the workers are setup to receive requests from the master, the worker_state or worker_init are not changed.

b) broadcast
Ok. Will test with order_tasks=True :)
